src/main/java/codeu/controller/ActivityServlet.java

```diff
@@ -0,0 +1,113 @@
+   */
+  void setMessageStore(MessageStore messageStore) {
+    this.messageStore = messageStore;
+  }
+
+  /**
+   * Sets the UserStore used by this servlet. This function provides a common setup method for use
+   * by the test framework or the servlet's init() function.
+   */
+  void setUserStore(UserStore userStore) {
+    this.userStore = userStore;
+  }
+
+  /**
+   * This function fires when a user navigates to the Activity page. It gets the conversations, messages,
+	 * and users from their respective stores and sorts them into one list by the most recent creation time.
+	 * It then forwards that list to activity.jsp in order to render it.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response)
+      throws IOException, ServletException {
+    List<Conversation> conversations = conversationStore.getAllConversations();
+    List<Message> messages = messageStore.getAllMessages();
+    List<User> users = userStore.getAllUsers();
+
+    List<Object> activity = new ArrayList<>();
+
+    int c = conversations.size() - 1;
+    int m = messages.size() - 1;
+    int u = users.size() - 1;
+
+    while ((c >= 0) || (m >= 0) || (u >= 0)) {
+      if ((c == -1 && u == -1) ||
+          (c == -1 && m != -1 && u != -1 && messages.get(m).getCreationTime().isAfter(users.get(u).getCreationTime())) ||
+          (u == -1 && m != -1 && c != -1 && messages.get(m).getCreationTime().isAfter(conversations.get(c).getCreationTime())) ||
+          (c != -1 && m != -1 && u != -1 && messages.get(m).getCreationTime().isAfter(conversations.get(c).getCreationTime()) &&
+             messages.get(m).getCreationTime().isAfter(users.get(u).getCreationTime()))) {
+        activity.add(messages.get(m));
+        m--;
+      } else if ((c == -1 && m == -1) ||
+       	         (c == -1 && u != -1 && m != -1 && users.get(u).getCreationTime().isAfter(messages.get(m).getCreationTime())) ||
+                 (m == -1 && u != -1 && c != -1 && users.get(u).getCreationTime().isAfter(conversations.get(c).getCreationTime())) ||
+                 (c != -1 && m != -1 && u != -1 && users.get(u).getCreationTime().isAfter(conversations.get(c).getCreationTime()) &&
+                    users.get(u).getCreationTime().isAfter(messages.get(m).getCreationTime()))) {
+        activity.add(users.get(u));
+        u--;
+      } else if ((u == -1 && m == -1) ||
+                 (u == -1 && c != -1 && m != -1 && conversations.get(c).getCreationTime().isAfter(messages.get(m).getCreationTime())) ||
+                 (m == -1 && c != -1 && u != -1 && conversations.get(c).getCreationTime().isAfter(users.get(u).getCreationTime())) ||
+                 (c != -1 && m != -1 && u != -1 && conversations.get(c).getCreationTime().isAfter(users.get(u).getCreationTime()) &&
+                    conversations.get(c).getCreationTime().isAfter(messages.get(m).getCreationTime()))) {
+        activity.add(conversations.get(c));
```

Comment:
I'm thinking making an ActivityStore list and adding the object (conversation, user, message) to the list at the same time it's being created may be a better idea. Alternatively, let's think of a more efficient sorting algorithm to sort this out

Reply:
Do you mean making ActivityStore a super class/ abstract class that Conversation, User, Message extend? If so, I agree. You can also just define a comparator method in the class and use sort. No need to reinvent the wheel. 👌